tensorflow/core/kernels/dml_conv_ops.cc

```diff
@@ -1347,4 +1348,598 @@ class DmlDepthwiseConv2DBackpropInputKernel : public DmlKernel {
 TF_CALL_DML_FLOAT_TYPES(DML_REGISTER_KERNEL);
 #undef DML_REGISTER_KERNEL
 
+template <bool HasDataFormatAttribute>
+struct Conv3DAttributes {
+  explicit Conv3DAttributes(OpKernelConstruction* context) {
+    if constexpr (HasDataFormatAttribute) {
+      string data_format_str;
+      OP_REQUIRES_OK(context,
+                     context->GetAttr("data_format", &data_format_str));
+      OP_REQUIRES(context, FormatFromString(data_format_str, &data_format),
+                  errors::InvalidArgument("Invalid data format"));
+    } else {
+      // V2 conv3d grad ops have a data format attribute but the original ops
+      // assume NHWC format.
+      data_format = FORMAT_NHWC;
+    }
+
+    std::vector<int32> strides;
+    OP_REQUIRES_OK(context, context->GetAttr("strides", &strides));
+    OP_REQUIRES(context, strides.size() == 5,
+                errors::InvalidArgument("Sliding window strides field must "
+                                        "specify 5 dimensions"));
+
+    int32 stride_n = GetTensorDim(strides, data_format, 'N');
+    int32 stride_c = GetTensorDim(strides, data_format, 'C');
+    stride_d = GetTensorDim(strides, data_format, '0');
+    stride_h = GetTensorDim(strides, data_format, '1');
+    stride_w = GetTensorDim(strides, data_format, '2');
+
+    OP_REQUIRES(
+        context, (stride_n == 1 && stride_c == 1),
+        errors::InvalidArgument("Current implementation does not yet support "
+                                "strides in the batch and depth dimensions."));
+    OP_REQUIRES(
+        context, (stride_d > 0 && stride_h > 0 && stride_w > 0),
+        errors::InvalidArgument("Spatial strides should be larger than 0."));
+
+    std::vector<int32> dilations;
+    OP_REQUIRES_OK(context, context->GetAttr("dilations", &dilations));
+    OP_REQUIRES(context, dilations.size() == 5,
+                errors::InvalidArgument("Dilation rates field must "
+                                        "specify 5 dimensions"));
+
+    int32 dilation_n = GetTensorDim(dilations, data_format, 'N');
+    int32 dilation_c = GetTensorDim(dilations, data_format, 'C');
+    dilation_d = GetTensorDim(dilations, data_format, '0');
+    dilation_h = GetTensorDim(dilations, data_format, '1');
+    dilation_w = GetTensorDim(dilations, data_format, '2');
+
+    OP_REQUIRES(context, (dilation_n == 1 && dilation_c == 1),
+                errors::InvalidArgument(
+                    "Current implementation does not yet support "
+                    "dilation rates in the batch and depth dimensions."));
```

Comment:
Comment says `depth`, but test says `dilation_c`. Guessing you meant `channel` instead.

Reply:
TF actually refers to the channels as depth in this context, so we want our error messages to be consistent with theirs -- it's not a typo. They don't even name the depth dimension in 3D conv: it's just spatial dim 0, spatial dim 1, spatial dim 2... that's why they can reuse the name depth. :(